Documentation and model of SNAP8-ER (#138)
idaholab/virtual_test_bed: doc/content/microreactors/s8er/s8er_multiphysics_model.md

Some aims of this model was to reduce the use of resources external to MOOSE, i.e. Cubit and to simplify the workflow from cross section generation to full griffin model. For this reason, the mesh is fully generated inside MOOSE using a variety of [Mesh Generator Objects](https://mooseframework.inl.gov/syntax/Mesh/index.html). Also for this reason, the [DFEM-SN Transport Scheme](https://griffin-docs.hpcondemand.inl.gov/latest/syntax/DFEM-SN/index.html), rather than the [CFEM-Diffusion Scheme](https://griffin-docs.hpcondemand.inl.gov/latest/syntax/CFEM-Diffusion/index.html) is used in the neutronic model. This allows us to have more a direct workflow from generation of the few group parameters and our griffin model, such as avoiding the generation of SPH correction factors, often needed for good agreement between griffin and reference solutions when using CFEM-Diffusion. DFEM-SN is also heavilyy optimized for parallel computation and comes in handy for larger problems such as this. 

!alert note
To request access to Bison or Griffin, please submit a request on the
[INL Modeling and Software Website](https://inl.gov/ncrc/).

## Generation of Few Group Parameters

> For generation of the reference solution and neutronic data the monte carlo code [Serpent](https://serpent.vtt.fi/mediawiki/index.php/Main_Page) was used. It must be noted that this model is quite sensitive to the energy structure and spatial resolution used in the generation of the few group parameters. Various sensiitivty studies were conducted to find the ideal parameters that were conducive to good agreement between Griffin and Serpent.

Reviewer: ```suggestion
For generation of the reference solution and neutronic data the monte carlo code [Serpent](https://serpent.vtt.fi/mediawiki/index.php/Main_Page) was used. It must be noted that this model is quite sensitive to the energy structure and spatial resolution used in the generation of the few group parameters. Various sensitivity studies were conducted to find the ideal parameters that were conducive to good agreement between Griffin and Serpent.
```

Reference data at each state point are generated for each tabulation parameter: fuel and coolant temperature, and each tabulation point: 300K, 600K, 900K, 1200K. This leads to a total of $2^{4}$ state points. The reflector temperature and the coolant density are maintained constant for each statepoint. Future updates to the model will include these effects. 

!table id=Serpent Model State Points caption= Serpent Model State Points
| Case      | Grid Index  | Fuel Temperature (K) | Coolant Temperature (K) |
| ----------- | ----------- |----------- | ----------- |
| 1      |   1 1     | 300 | 300 |
| 2      |   1 2     | 300 | 600 |
| 3      |   1 3     | 300 | 900 |
| 4      |   1 4     | 300 | 1200 |
| 5      |   2 1     | 600 | 300 |
| 6      |   2 2     | 600 | 600 |
| 7      |   2 3     | 600 | 900 |
| 8      |   2 4     | 600 | 1200 |
| 9      |   3 1     | 900 | 300 |
| 10     |   3 2     | 900 | 600 |
| 11     |   3 3     | 900 | 900 |
| 12     |   3 4     | 900 | 1200 |
| 13     |   4 1     | 1200 | 300 |
| 14     |   4 2     | 1200 | 600 |
| 15     |   4 3     | 1200 | 900 |
| 16     |   4 4     | 1200 | 1200 |

### Preparation of Neutronic Data for Griffin

> Now having all the reference neutronic data generated by serpent, the data must be prepared for use in our griffin model. This is done through [ISOXML](https://griffin-docs.hpcondemand.inl.gov/latest/isoxml-user-manual/introduction.html#introduction), a general toolkit aimed at handling large multigroup libraries. This is also comes in handy for faciliating the coupling process and handling feedback effects by using the data generated for our various state points. 

Reviewer: ```suggestion
Now having all the reference neutronic data generated by serpent, the data must be prepared for use in our griffin model. This is done through [ISOXML](https://griffin-docs.hpcondemand.inl.gov/latest/isoxml-user-manual/introduction.html#introduction), a general toolkit aimed at handling large multigroup libraries. This is also comes in handy for facilitating the coupling process and handling feedback effects by using the data generated for our various state points. 
```

!listing microreactors/s8er/core_2D_griffin_coupled.i block=Mesh

1. 9 individual fuel channels are modeled, 1 for each corresponding radial ring 
2. The fuel channels are mapped into the lattice in the appropiate arrangement from the homogenized universe data.
3. A peripheral ring is enclosed around the lattice to represent the internal reflector + core barrel (last radial layer)
4. The mesh is cut along the ring at various points to preserve the geometry and dimension of the reference serpent model. 

> 5. The blocks are mapped to their appropiate material ID.

Reviewer: ```suggestion
5. The blocks are mapped to their appropriate material ID.
```

A table with corresponding mapping between the Serpent Universe ID and the Griffin Block ID is shown below.

!table id=Homogenized Universe Mapping caption= Homogenized Universe Mapping
| Serpent Universe ID | Material ID | Block ID | Material/Region |
| -------- | ------- | ----- | ------------ |
| 100      |   100   |  1    | Fuel Channel |
| 200      |   200   |  2    | Fuel Channel |
| 900      |   900   |  3    | Fuel Channel |
| 400      |   400   |  4    | Fuel Channel |
| 500      |   500   |  5    | Fuel Channel |
| 600      |   600   |  6    | Fuel Channel |
| 700      |   700   |  7    | Fuel Channel |
| 800      |   800   |  8    | Fuel Channel |
| 300      |   300   |  9, 10 | Coolant Channel |
| 1200     |   1200  |  11   | Internal Reflector + Core Barrel |

!media s8er/griffin_universe_map.png
  caption= S8ER Griffin Mesh, from [!citep](s8er_naupa2022)
  style=width:100%;margin-left:auto;margin-right:auto

### TransportSystems

> As stated previously, the Griffin model uses the newly optimized DFEM-SN transport scheme to solve for the eigenvalue. Here a 16 group energy structure is used with vacuum boundary conditions on the outeredge of the core barrel. The solver uses 9 azimuthal and 3 polar angles, typically the default for these are 6 and 2 respectively but because this highly anistropic, a higher order of angles is required for solution fidelity. The benefit with using the DFEM-SN scheme rather than the stand CFEM-Diffusion scheme, is the removal of several layers of calculation for the same fidelity, however at the expense of a more expensive computation. The DFEM-SN scheme is optimized and scales well with problem size. 

Reviewer: ```suggestion
As stated previously, the Griffin model uses the newly optimized DFEM-SN transport scheme to solve for the eigenvalue. Here a 16 group energy structure is used with vacuum boundary conditions on the outer edge of the core barrel. The solver uses 9 azimuthal and 3 polar angles, typically the default for these are 6 and 2 respectively but because this system is highly anisotropic, a higher order of angles is required for solution fidelity. The benefit with using the DFEM-SN scheme rather than the stand CFEM-Diffusion scheme, is the removal of several layers of calculation for the same fidelity, however at the expense of a more expensive computation. The DFEM-SN scheme is optimized and scales well with problem size. 
```

Reviewer: maybe per quarter? it s likely 6 polar in 3D not 3, and 36 azimuthal?

!listing microreactors/s8er/core_2D_griffin_coupled.i block=TransportSystems

### Materials

> Through the use of the [CoupledFeedbackMatIDNeutronicsMaterial](https://griffin-docs.hpcondemand.inl.gov/latest/source/materials/CoupledFeedbackMatIDNeutronicsMaterial.html#coupledfeedbackmatidneutronicsmaterial), by the mapping of `materialIDs` to the mesh done earlier, this object conviniently handles the direct mapping of neutronic data into the mesh. The object also has a feature to correct the neutronic data used in a material by adjusting the cross sections for a correction volume. This is done for the internal reflector and core barrle region to more appropiately match the reference model.

Reviewer: ```suggestion
Through the use of the [CoupledFeedbackMatIDNeutronicsMaterial](https://griffin-docs.hpcondemand.inl.gov/latest/source/materials/CoupledFeedbackMatIDNeutronicsMaterial.html#coupledfeedbackmatidneutronicsmaterial), by the mapping of `materialIDs` to the mesh done earlier, this object conveniently handles the direct mapping of neutronic data into the mesh. The object also has a feature to correct the neutronic data used in a material by adjusting the cross sections for a correction volume. This is done for the internal reflector and core barrel region to more appropriately match the reference model.
```

!listing microreactors/s8er/core_2D_griffin_coupled.i block=Materials

## Bison Model

> Because bison is only taking the power distribution from the griffin model, we have the freedom to model each fuel channel component explciitly and hence we are able to capture unique heat transfer phenomena such as the gap radiative transfer and model thermal properties for each component. This model only captures heat conduction but future updates will include feedback from thermomechanical effects such as thermal expansion. 

Reviewer: ```suggestion
Because bison is only taking the power distribution from the griffin model, we have the freedom to model each fuel channel component explicitly and hence we are able to capture unique heat transfer phenomena such as the gap radiative transfer and model thermal properties for each component. This model only captures heat conduction but future updates will include feedback from thermomechanical effects such as thermal expansion. 
```

!media s8er/fuelpin_components_label_bison.png
  caption= S8ER Bison Mesh, from [!citep](s8er_naupa2022)
  style=width:100%;margin-left:auto;margin-right:auto

### Geometry 

> The process for generating the Bison mesh is almost identical to the process described for the Griffin mesh, with a few extra steps. This time since were modeling the explicit fuel element geometry, we use the [PolygonConcentricCircleMeshGenerator](https://mooseframework.inl.gov/source/meshgenerators/PolygonConcentricCircleMeshGenerator.html) to model concentric rings correpsondig to the outer radii of each material in the fuel element.

Reviewer: ```suggestion
The process for generating the Bison mesh is almost identical to the process described for the Griffin mesh, with a few extra steps. This time since we are modeling the explicit fuel element geometry, we use the [PolygonConcentricCircleMeshGenerator](https://mooseframework.inl.gov/source/meshgenerators/PolygonConcentricCircleMeshGenerator.html) to model concentric rings corresponding to the outer radii of each material in the fuel element.
```

!listing microreactors/s8er/core_2D_bison_coupled.i block=Mesh

!table id=S8ER Bison Dimensions caption= S8ER Bison Dimensions
| Material/Region |  Radius (m) |
| --------------- | --------- |
| UZrH Radius              |   0.0067564)   |
| Ceramic Outer Radius     |   0.007130879  |
| Gap Outer Radius         |   0.006866719  |
| Cladding Outer Radius    |   0.007130879  |
| Lattice Pitch            |   0.007239     |

### Kernels, AuxKernels, Functions

> The Kernels used are the [HeatConduction](https://mooseframework.inl.gov/source/kernels/ADHeatConduction.html) to solve the steady state heat conduction equation, aswell as the [CoupledForce](https://mooseframework.inl.gov/source/kernels/CoupledForce.html) to couple the power distribution from bison as the spatially dependent heat generation term.

Reviewer: ```suggestion
The Kernels used are the [HeatConduction](https://mooseframework.inl.gov/source/kernels/ADHeatConduction.html) to solve the steady state heat conduction equation, as well as the [CoupledForce](https://mooseframework.inl.gov/source/kernels/CoupledForce.html) to couple the power distribution from bison as the spatially dependent heat generation term.
```

!listing microreactors/s8er/core_2D_bison_coupled.i block=Kernels

The AuxKernels used are a variety of function executors and variable normalizers for solving thermal properties that are temperature dependent, normalizing the power distribution for power conservation between Bison and Griffin, and for seperating the fuel, coolant, and reflector temperatures into individual variables to be used as the tabulation parameters.

!listing microreactors/s8er/core_2D_bison_coupled.i block=AuxKernels

The Functions used include the thermal conductivity, and specific heat of the fuel as a function of temperature from [!citep](SNAPFuel). 

!listing microreactors/s8er/core_2D_bison_coupled.i block=Functions

### Materials

> The thermal properties of each material are modeled through using [GenericConstantMaterial](https://mooseframework.inl.gov/source/materials/GenericConstantMaterial.html). This allows you to set thermal properties such as the density, specifc heat,  and thermal conductivity. In the case of the fuel we have documented formulations for the specific heat and thermal conductivity as a function of temperature. This temperature dependence can implemented through the use of the [HeatConductionMaterial](https://mooseframework.inl.gov/source/materials/HeatConductionMaterial.html) which allows you to set material properties as a function of temperature. 

Reviewer: ```suggestion
The thermal properties of each material are modeled through using [GenericConstantMaterial](https://mooseframework.inl.gov/source/materials/GenericConstantMaterial.html). This allows you to set thermal properties such as the density, specific heat,  and thermal conductivity. In the case of the fuel we have documented formulations for the specific heat and thermal conductivity as a function of temperature. This temperature dependence can implemented through the use of the [HeatConductionMaterial](https://mooseframework.inl.gov/source/materials/HeatConductionMaterial.html) which allows you to set material properties as a function of temperature. 
```

The temperature distribution for each respective tabulation parameter i.e, fuel temperature and coolant temperature is transfered using the [MultiAppInterpolationTransfer](https://mooseframework.inl.gov/source/transfers/MultiAppInterpolationTransfer.html). This transfer is suitable for nodal variables such as variables in the `LAGRANGE` family.   

!listing microreactors/s8er/core_2D_griffin_coupled.i block=Transfers

> Reffering back to the `[Materials]` block, this is where the coupled feedback is taken into account. The [CoupledFeedbackMatIDNeutronicsMaterial](https://griffin-docs.hpcondemand.inl.gov/latest/source/materials/CoupledFeedbackMatIDNeutronicsMaterial.html#coupledfeedbackmatidneutronicsmaterial) takes the tabulated neutronic data and uses temperature from bison to interpolate between statepoints. For this reason,  Mutliple transfers for each tabulation parameter are required so that the nuetronic data can be properly interpolated between statepoints.

Reviewer: ```suggestion
Referring back to the `[Materials]` block, this is where the coupled feedback is taken into account. The [CoupledFeedbackMatIDNeutronicsMaterial](https://griffin-docs.hpcondemand.inl.gov/latest/source/materials/CoupledFeedbackMatIDNeutronicsMaterial.html#coupledfeedbackmatidneutronicsmaterial) takes the tabulated neutronic data and uses temperature from bison to interpolate between statepoints. For this reason,  Multiple transfers for each tabulation parameter are required so that the neutronic data can be properly interpolated between statepoints.
```
